### PR TITLE
Rperform R CMD Check

### PR DESCRIPTION
**Pull request by EngineerDanny**
This PR creates a Github Action for setting up R, Rperform and its dependencies on the remote machine. 
The test is carried out on the [stringr](https://github.com/tidyverse/stringr) package.

The minimal test is to run Rperform on this package successfully.

### CODE REVIEW

**Comment by EngineerDanny** [comment]
**Implementation Review**

The main focus is to be able to run Rperform on **other** packages(eg. [stringr](https://github.com/tdhock/stringr) ).
On this repository, there will be a minimal testing of the package itself.
The Github Action tests will be done on the repositories of other packages which I will fork. 
For conveniences, I will proceed the testing with the stringr package [here](https://github.com/EngineerDanny/stringr).

**Comment by EngineerDanny** [comment]
The Rperform package has been failing R CMD checks.

Screenshot 2022-06-22 at 11 44 48 AM

As it can be seen, all the tests passed except running `r-lib/actions/check-r-package@v2`
After running the package check locally, even though the test didn't explicitly give me an error, it failed.

The only warning was : 

>   checking Rd cross-references ... WARNING
>   Missing link or links in documentation object 'Rperform-package.Rd':
>     ‘[git2r:git2r-package]{git2r}’
>   
>   See section 'Cross-references' in the 'Writing R Extensions' manual.

I fixed the warning by inserting the [git2r](https://github.com/ropensci/git2r) official github link.

**Comment by EngineerDanny** [comment]
Initially, the issue was invalid link in the `test-repo-metrics.R `file.
I found out that the [hadley/stringr](https://github.com/hadley/stringr) link does not exist.
Even though, when you open the link you will be redirected to the actual repository which is the [tidyverse/stringr](https://github.com/tidyverse/stringr) .

My test results were inconsistent (Failing more times and occassionally passing) due to this issue.
Therefore, I changed the github link to [tidyverse/stringr](https://github.com/tidyverse/stringr)

**Comment by EngineerDanny** [comment]
Although the issue with the link was fixed and **R CMD Check** on my local machine passed, the GitHub CI action still failed.
So I started to dig into the [tidyverse/stringr](https://github.com/tidyverse/stringr) package tests.

Inside the https://github.com/tidyverse/stringr/blob/main/tests/testthat/test-dup.r file, there are three tests which are ran internally. 

Screenshot 2022-06-22 at 2 30 21 PM

As shown above, all the other tests passed except the `uses tidyverse recycling rules` test.
This led to my conclusion that, this may be an internal issue with the `tidyverse/stringr` package.

The first two tests however, could be used in checking Rperform.

**Comment by EngineerDanny** [comment]
After several debugging and testing, I finally found the final error that was causing the **Rperform R CMD Check** test to fail on GitHub CI. It turns out that during my adventures and explorations, I mistakenly created a file on the master branch   
👉 [tests/testthat/test-plot-metrics.R]. 

Screenshot 2022-06-22 at 8 14 21 PM

As recorded from the logs above, in the file, there is an issue with the `str_c`  function because it does not exist. 
Therefore, I have removed this file since it's not needed.

Now, all checks have passed ✅.

File: R/branch_metrics.R [hunk 1]
```diff
@@ -256,7 +256,7 @@ compare_branchm <- function(test_path, branch1, branch2 = "master") {
   target2 <- git2r::repository(file.path("./"))
   # If branch2 is specified, check out to it and obtain commit list
   if (!is.null(branch2)) {
-    original_state2 <- git2r::head(target2)
+    original_state2 <- git2r::repository_head(target2)
```
Comment (tdhock) [inline R/branch_metrics.R:259]: you have these changes in 3 different branches/prs, right? it is better for review to keep them in just one branch/pr so your reviewers (me) only have to review the change once.

**Comment by tdhock** [comment]
also i'm not sure if this is the case for the stringr that you are using, but it would be good to have a test case in which there is a clear slowdown/speedup that can be observed, for example this branch/pr probly has a slow down  https://github.com/tdhock/binsegRcpp/pull/10

File: R/branch_metrics.R (hunk 1 above)
Comment (EngineerDanny) [reply R/branch_metrics.R:259]: Sure, I will keep all the changes in **test-pkg** branch. I will also keep only this PR open.

**Comment by EngineerDanny** [comment]
My new commits to this repo fix most of the issues raised.

1. Indentation inconsistency in the R files have been fixed by using R studio's built-in indentation. [(1)](https://github.com/EngineerDanny/Rperform/pull/2#discussion_r904718686)
2. Various binary files and directories have been added to the **.gitignore** file. [(2)](https://github.com/EngineerDanny/Rperform/pull/2#issuecomment-1164099213)
3. Named reference approach has been adopted to reference list members by names instead of numeric indexes. [(3)](https://github.com/EngineerDanny/Rperform/pull/3#discussion_r904711183)

**Comment by EngineerDanny** [comment]
@tdhock My recent tests on [PR/branch](https://github.com/tdhock/binsegRcpp/pull/10) with Rperform show that your assumption is right.
Merging this [PR](https://github.com/tdhock/binsegRcpp/pull/10) made the code consistently slower.

Since they were on different branches, I ran the 
`plot_branchmetrics(test_path = "tests/testthat/test-CRAN.R", metric = "time", branch1 = "l1loss", branch2 = "master", save_data = T, save_plots = F)` function and observed the results.

I noticed that, the result was not depicted clearly on the graph. #4 
However, the **time_frame** data set that was saved had 254 observations and 8 variables.
This is the data 👉 [time_frame.csv](https://github.com/EngineerDanny/Rperform/files/8969421/time_frame.csv)

**Comment by EngineerDanny** [comment]
**BREAKING DOWN THE RESULTS**
The results were consistent. After the [PR](https://github.com/tdhock/binsegRcpp/pull/10) was merged, there was a clear slowdown that was observed.

**1. The first test was "one data point has zero loss"**
Screenshot 2022-06-23 at 4 34 30 PM

**Average Time Taken**
Before PR : 0.009315 seconds. 
After merging PR : 0.0180096 seconds (Approximately 2 times increase)

**2. "equal split cost is ok" test** 
Screenshot 2022-06-23 at 4 35 27 PM

**Average Time Taken**
Before PR : 0.0351191 seconds. 
After merging PR : 0.057926 seconds (Approximately 1.6 times increase)

**3. "error for 0 data" test**
Screenshot 2022-06-23 at 4 36 09 PM

**Average Time Taken**
Before PR : 0.0072106 seconds. 
After merging PR : 0.0352905 seconds (Approximately 5 times increase)

The list goes on and on, showing a consistent increase in time after merging the PR.

**Pull request merged**